main/src/main/scala/sbt/internal/CommandExchange.scala

```diff
@@ -140,7 +142,18 @@ private[sbt] final class CommandExchange {
       val portfile = s.baseDir / "project" / "target" / "active.json"
       val h = Hash.halfHashString(IO.toURI(portfile).toString)
       val tokenfile = BuildPaths.getGlobalBase(s) / "server" / h / "token.json"
-      val socketfile = BuildPaths.getGlobalBase(s) / "server" / h / "sock"
+      val socketfile = {
+        val socketfile = BuildPaths.getGlobalBase(s) / "server" / h / "sock"
+        connectionType match {
+          case ConnectionType.Local if !Util.isWindows =>
+            val maxSocketLength = new UnixDomainSocketLibrary.SockaddrUn().sunPath.length - 1
+            if (socketfile.absolutePath.length > maxSocketLength)
+              Files.createTempFile("sbt-server", ".sock").toFile // assuming this is short enough..
```

Comment:
Assuming that there might be multiple independent sbt instances running at the same time, in which case you want the `.sock` file to be unique, wouldn't it be better to just always hash the string? Something like:
```
Files.createTempFile("sbt-server-"+socketfile.getCanonicalPath.hashCode.toString, ".sock").toFile
```
or, even better:
```
import java.security.MessageDigest;

MessageDigest digest = MessageDigest.getInstance("SHA-256");
d.update(socketfile.getCanonicalPath.getBytes());
Files.createTempFile("sbt-server-"+messageDigest.digest().toString, ".sock").toFile
```
Please also note that `getCanonicalPath` is preferable to the absolute path, since the latter is not unique (symbolic links are not resolved.)

Reply:
> Assuming that there might be multiple independent sbt instances running at the same time, in which case you want the .sock file to be unique, wouldn't it be better to just always hash the string?

We do want a unique file: that's what `Files.createTempFile` does:

Reply:
Right, ok. Then why not just using "sbt-server" as a base, regardless of the dir path?...

Reply:
Isn't that what I'm doing? Using just "sbt-server" as a base regardless of the dir path?

Reply:
Do you mean get rid of the ".sock" suffix and use `null` instead?

Reply:
No, I mean: I see that in the revised code you are using the dir name as a base for the socket file, unless the path is too long, in which case you use "sbt-server". It there any specific advantage in using the path name, instead of always using just the "sbt-server" string in all cases? It's not very important, just curious.

Reply:
I'm not sure what the original design decisions were for having the socket file in the global sbt directory. I'm just leaving that code path as is. Perhaps @eed3si9n can enlighten us.

Reply:
There are two requirements
1. sock file should be somewhere the user can secure using chown and chmod. sbt global directory is a good candidate for that.
2. You can't have two server instances running on a build. That's why `h` is calculated using the hash of portfile path.

Reply:
So what do you recommend if the sbt global directory's path is too long to use as the path of the socket?

Reply:
Maybe introduce a environment variable the user can use to specify "server" directory, and fail to start the server with a message saying either change your global directory or set the env variable?

Reply:
ok

Reply:
Is there any reference on how to set this variable I believe somewhat in lines of `-Dsbt.global.server.directory'  ? I didnt find any reference on this except the release notes and this PR.

Reply:
Here's the first article I found on how to set an environment variable, if that's the question - https://www.twilio.com/en-us/blog/how-to-set-environment-variables-html